### PR TITLE
Toast rule

### PR DESCRIPTION
**Pull request by dsvoronin**

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -52,7 +52,7 @@ mockitoJunit5Version=2.23.4
 mockitoCoreVersion=2.18.3
 junit5PlatformVersion=1.6.0
 junit5Version=5.6.0
-junitVersion=4.12
+junitVersion=4.13
```
Comment (dsvoronin) [inline gradle.properties:56]: https://github.com/junit-team/junit4/blob/master/doc/ReleaseNotes4.13.md

File: docs/content/docs/test/Toast.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+---
+title: Testing toasts
+type: docs
+---
+
+# Testing toasts
```
Comment (sergioboi) [inline docs/content/docs/test/Toast.md:6]: Toast testing?

File: docs/content/docs/test/Toast.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+showToast("I'am a toast!")
+```
+
+```kotlin
+implementation("com.avito.android:proxy-toast:$version")
+```
+
+## ToastRule
+
+ToastRule changes proxy toast's singleton before the test to store `showToast` invocations.
```
Comment (sergioboi) [inline docs/content/docs/test/Toast.md]: Change abstract `changes`
```suggestion
ToastRule clears shown toast history that stores in proxy toast singleton before the test.
```

File: docs/content/docs/test/Toast.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+Use provided extension methods to be able to spy on `showToast` functions.
+
+```kotlin
+import com.avito.android.util.showToast
+
+showToast("I'am a toast!")
+```
+
+```kotlin
+implementation("com.avito.android:proxy-toast:$version")
```
Comment (sergioboi) [inline docs/content/docs/test/Toast.md:31]: Do we need describe all deps?

File: docs/content/docs/test/Toast.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+type: docs
+---
+
+# Testing toasts
+
+[android.widget.Toast](https://developer.android.com/reference/android/widget/Toast)
+
+{{< hint warning>}}
+Testing toasts reliably is a hard task because some other toast can show up and yours will be queued. 
+We hack app with proxy object to mitigate this.
```
Comment (sergioboi) [inline docs/content/docs/test/Toast.md]:
```suggestion
Testing toasts is a hard because other toasts can show up and yours will be queued. 
We hack app with proxy object to mitigate this.
```

File: settings.gradle.kts [hunk 6]
```diff
@@ -78,6 +78,9 @@ if (syncAndroidModules.toBoolean()) {
     include(":subprojects:android-test:test-report")
     include(":subprojects:android-test:test-inhouse-runner")
     include(":subprojects:android-test:test-app")
+    include(":subprojects:android-test:toast-rule")
```
Comment (eugene-krivobokov) [inline settings.gradle.kts:81]: I think that an independent module for toasts is too much.
We discussed an additional module with espresso specific code that can be used without our testing lib.

File: gradle.properties [hunk 7]
```diff
@@ -7,6 +7,7 @@ org.gradle.caching=true
 # https://issuetracker.google.com/issues/139438142
 # https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures
 syncAndroidModules=true
+android.enableR8=true
```
Comment (eugene-krivobokov) [inline gradle.properties:10]: It's better to make it in a separate PR

File: docs/content/docs/test/Toast.md (hunk 3 above)
Comment (dsvoronin) [reply docs/content/docs/test/Toast.md]: i don't understand the phrase now

File: docs/content/docs/test/Toast.md (hunk 2 above)
Comment (dsvoronin) [reply docs/content/docs/test/Toast.md:6]: original one looks better to me

File: docs/content/docs/test/Toast.md (hunk 4 above)
Comment (dsvoronin) [reply docs/content/docs/test/Toast.md:31]: what deps? it's enough for this library to run

File: docs/content/docs/test/Toast.md (hunk 3 above)
Comment (sergioboi) [reply docs/content/docs/test/Toast.md]: I've updated

File: docs/content/docs/test/Toast.md (hunk 4 above)
Comment (sergioboi) [reply docs/content/docs/test/Toast.md:31]: I understand nothing.
You created two modules `proxy-toast` and `toast-rule`. You added them to test-app dependencies.
Is `proxy-toast` enough for what without toast-rule?
You described usages of ToastRule but didn't add description how to add it dependency. Comparing with `proxy-toast` why didn't you add dependency description?

Comment (dsvoronin) [reply docs/content/docs/test/Toast.md:31]: yep, i missed that, thanks

File: docs/content/docs/test/Toast.md (hunk 3 above)
Comment (dsvoronin) [reply docs/content/docs/test/Toast.md]: @eugene-krivobokov , @int02h, what do you think? which variant is better?

Comment (eugene-krivobokov) [reply docs/content/docs/test/Toast.md]: I'd prefer to avoid these implementation details here.
For me, the phrase "Use provided extension methods to be able to spy ... " was enough.  I don't need to know about singleton and history of invocations.

Comment (int02h) [reply docs/content/docs/test/Toast.md]: Maybe "Before the test ToastRule clears history of shown toasts stored in proxy toast singleton."?

File: gradle.properties (hunk 7 above)
Comment (int02h) [inline gradle.properties:10]: How does R8 relate to this pull request?

File: subprojects/android-test/toast-rule/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+plugins {
```
Comment (int02h) [inline subprojects/android-test/toast-rule/build.gradle.kts:1]: Do we really need a separate module just for toast rule?

Comment (dsvoronin) [reply subprojects/android-test/toast-rule/build.gradle.kts:1]: it's a weird hack, and i really don't want to ship it as a part of ui testing library
also i think it's a great idea to ship separate things(can be used without each other) separately

File: gradle.properties (hunk 7 above)
Comment (dsvoronin) [reply gradle.properties:10]: it's weird, but adding ToastTest triggered a bug in old version of keeper and i had to bump it, i found out that r8 was not enabled explicitly, even if it's by default now, so added it for clarity

File: docs/content/docs/test/Toast.md (hunk 3 above)
Comment (dsvoronin) [reply docs/content/docs/test/Toast.md]: it is not necessary to clear collection before test as we run all of them in isolation right now;
i don't want to focus on it here.
All i wanted to say that Rule changes singleton value to it's own Proxy object and keeps tracking of invocations. 
I think my original explanation good enough

Comment (eugene-krivobokov) [reply docs/content/docs/test/Toast.md]: >  Rule changes singleton value to it's own Proxy object and keeps tracking of invocations.

The more I read this the more I think that this is excessive implementation details.

File: settings.gradle.kts (hunk 6 above)
Comment (dsvoronin) [reply settings.gradle.kts]: removed it

Comment (dsvoronin) [reply settings.gradle.kts:81]: https://github.com/avito-tech/avito-android/pull/228#discussion_r388899533

File: subprojects/android-test/toast-rule/build.gradle.kts (hunk 8 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/build.gradle.kts:1]: I agree that this is not a part of the ui-testing library.
For me, it looks like a ktx analogy for Espresso.

**Review by eugene-krivobokov — APPROVED** [review]

File: docs/content/docs/test/Toast.md (hunk 3 above)
Comment (sergioboi) [reply docs/content/docs/test/Toast.md]: What is a value of toatrule info? Implementation details that teach how to use rule correctly give a value.

File: docs/content/docs/test/Toast.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+val toastRule = ToastRule()
+
+fun test() {
+    toastRule.checks.toastDisplayedWithText("I'am a toast!")
+}
+```
+
+```kotlin
+androidTestImplementation("com.avito.android:toast-rule:$version")
+```
```
Comment (sergioboi) [inline docs/content/docs/test/Toast.md:47]: It's look like abstract bull shit.
Point of this page isn't clear.
It's not clear that `showToast` must be used in production instead of standart functions.
It's not clear that Rule useless without `showToast`
Info about SnackBar is too large compare with other content.
Structure i see:
- You can't test toast
- We help you to test toast
- You can test toast that shows only by our `showToast` in production
- You can use our rule
- We recommend to use snackbar

Comment (eugene-krivobokov) [reply docs/content/docs/test/Toast.md:47]: > It's look like abstract bull shit.

I see your reaction and share some worries. But please, try to find other words. The problem is, it can be interpreted as aggression and doesn't provide information about exact problems.
https://avito-tech.github.io/avito-android/docs/contributing/codereview/#best-practices

Apart from this, I agree with the next proposals.

Comment (sergioboi) [reply docs/content/docs/test/Toast.md:47]: Sorry @dsvoronin . I want to say that text is hard to read. I will work on my messages

Comment (sergioboi) [reply docs/content/docs/test/Toast.md:47]: Do i need to create pr with proposals?
